File: 55. Jump Game/CanJump.py
'''
https://leetcode.com/problems/jump-game/
'''


from typing import List


class Solution:
    # O(n^2) time, 
    # O(n) space, 
    # Approach: Dynamic programming,
    # just made it past the test, needs optimization
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        if n == 1:
            return True
        dp = [False for i in range(n-1)]
        if nums[n-2] > 0:
            dp[n-2] = True
        
        for i in range(n-3, -1, -1):
            if (i + nums[i] >= n-1):
                dp[i] = True
                continue
            for j in range(1, nums[i]+1):
                if dp[i+j]:
                    dp[i] = True
                    break
        
        return dp[0]

File: 55. Jump Game/can_jump.rs
impl Solution {
    
    // O(n^2) time,
    // O(n) space, 
    // Approach: dp, bottom up, 
    pub fn can_jump(nums: Vec<i32>) -> bool {
        let mut dp: Vec<i32> = vec![0; nums.len()];
        dp[nums.len()-1] = 1;
        
        for i in (0..nums.len()-1).rev() {
            let jumps: usize = nums[i] as usize;
            let mut found = false;
            for jump in 1..jumps+1 {
                if i+jump < nums.len() && dp[i+jump] == 1 {
                    found = true;
                    break;
                }
            }
            if found {
                dp[i] = 1;
            }
            
        }
        return dp[0] == 1;
    }

    
    // O(n) time,
    // O(1) space, 
    // Approach: greedy, 
    pub fn can_jump(nums: Vec<i32>) -> bool {
        let mut farthest: i32 = 0;
        
        for i in 0..nums.len() {
            let jump: i32 = nums[i];
            if i as i32 <= farthest {
                farthest = cmp::max(farthest, i as i32+jump);
            } else {
                return false
            }
        }
        
        return true;
    }
}